### ISSUE TITLE
easier command prompt tab completion

### ISSUE THREAD

**Issue by aeosynth** [martanne/vis#526]
request: automatic filename tab completion in the command prompt. currently this must be activated with eg `:e .`; i'd like to write `:e f<Tab>` and have it expand to `:e foo`.

**Comment by josuah** [comment]
See #100 as well.

In command mode, the input could be piped through an external script, managing the completion on its own, and piping it back to vis once fully completed.

`vis-menu` could be used multiple times in the script:

1. If tab is pressed, the command is selected with vis-menu,
2. Another vis-menu prompt is called depending on the command used to select the arguments.  The previous input can be displayed as a prompt with the `-p` flag:

If tab is being pressed in command-mode, start the script with vis-menu:

```
[:]                   {w} | e | wq | cd | x | ...
```
It prompts for some command.
```
[:] e                 {e}
``` 
Enter is pressed, the result is piped to the script.  The command is "e" -> start vis-open
```
[:e]                  {.} | .. | src | doc | bin | etc | ...
```
```
[:e] s               {src} | downloads | ...
```
```
[:e src]              vis | {vis-menu} | other-project | ...
```

```
[:e src/vis-menu]     {vis-menu.c}
```
<kbd>Enter</kbd>, and the script is done.  The full string produced will be: `:e src/vis-menu/vis-menu.c`, which would come back to vis.

This approach is used by [sandy(1)](http://tools.suckless.org/sandy), useful for the editor-side implementation:

- http://git.suckless.org/sandy/tree/config.def.h#n33
- http://git.suckless.org/sandy/tree/config.def.h#n40
- http://git.suckless.org/sandy/tree/config.def.h#n143
- http://git.suckless.org/sandy/tree/sandy.c#n582

And by @halfwit's [dsearch(1)](https://github.com/halfwit/dsearch) scripts for dmenu, with a set of `!command`s, useful for the script-side implementation:

- https://github.com/halfwit/dsearch/blob/master/dsearch#L33

**Comment by martanne** [comment]
@josuah I don't think this is really needed for the few single letter commands. I guess it would improve discoverability, but running commands without knowing what they are doing seem like a bad idea anyway. Just read the man page or the `:help :-Commands` output for that.

@aeosynth Assuming that `vis-complete` would immediately return+terminate in case only one completion candidate is available (as `vis-open` already does). What you want could mostly be achieved by a window local insert mode mapping `<Tab>` to `<C-x><C-f>` (file name completion)?. @Screwtapello mentioned that given enough time, he might be able to make the necessary `vis-complete` changes.

**Comment by aeosynth** [comment]
`<C-x><C-f>` works, but the ui is not as good as `vis-open`, since the former shows full paths, and the latter shows just the filenames. `<C-x><C-f>` is almost unusable for drilling down multiple directories. i sometimes open up vim just for better filename completions.

**Comment by josuah** [comment]
@martanne Well...  This may have been a little bit too far on completion.  And having vis keybindings is still great.

If someone is still interested, this may be implemented as a plugin I guess, as it is filtering a line through an external shell script.  No change to be done for vis for this to work!

**Comment by martanne** [comment]
@aeosynth I guess you could use a mapping `<Tab>` -> `*<Enter>`, but that can potentially be very dangerous.
